### DISCUSSION TITLE
Convergence issues due to huge differences in stiffness tensor

### DISCUSSION THREAD

**Discussion by Bala-1005** [idaholab/moose#16922] [Q&A Modules: Solid mechanics] [answered]
Hello everyone, 

I am working on kks model for solidification with the two phases being solid and liquid. I have included the tensor mechanics module too similar to the one shown in the example "/moose/modules/combined/examples/phase_field-mechanics/kks_mechanics_VTS.i". 

The stiffness of the liquid phase is several orders lower than the solid(~10^9). This has caused some convergence issues close to the boundary and when two solid parts of the domain interact. I am not sure how to alleviate this problem as this seems to be the only issue with the code. I have fiddled around with different boundary conditions but that has not helped at all.

How do I solve this issue?

Thanks,
Bala

**Comment by aeslaughter** [comment, marked as answer]
Have you tried using automatic_scaling? This can be enabled in the Executioner block with "automatic_scaling = true"

**Reply by Bala-1005** [reply]
I did try that but it doesn't seem to help. I also tried to use ActivateElementsCoupled as instructed in #16715(This is a related issue where I wanted to activate tensor mechanics only for the solid and interface part but not the liquid phase). But it gives me the following error.

"A 'ActivateElementsCoupled' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable."


Thanks,
Bala

**Reply by GiudGiud** [reply]
Did you build your application with tensor_mechanics? It says what module is included in the Makefile

**Reply by Bala-1005** [reply]
Yes. I did build my application with tensor_mechanics. Besides, I had the physics wrong when I wanted to activate only the solid part of the domain for tensor mechanics. Right now I have applied tensor mechanics to the entire domain. Convergence is still an issue at the boundary.

**Reply by jiangwen84** [reply]
If you are using phase-field variable to indicate the phase domain, you should not use ActivateElementsCoupled. That will assign phase domain element-wise.

Is your simulation running fine with two comparable stiffness tensors? Are you using a direct solver LU in the executioner block?

**Reply by Bala-1005** [reply]
Yes. The simulation runs fine with comparable stiffness tensors. I was able to solve the issue at the boundary by deactivating some of the boundary conditions. But when two solids meet, the simulation ceases to converge. I have attached an image
 at the point where the simulation stops.

![output2](https://user-images.githubusercontent.com/75709531/111533902-b6873080-873d-11eb-9dca-40d0ea217cad.png)

Also, here is the executioner block.
[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type'
  petsc_options_value = 'asm      ilu          nonzero'
  l_max_its = 30
  nl_max_its = 50
  nl_rel_tol = 1e-08
  automatic_scaling = true
  end_time = 25
  dt = 0.008
 [./Adaptivity]
    initial_adaptivity = 3 # Number of times mesh is adapted to initial condition
    refine_fraction = 0.6 # Fraction of high error that will be refined
    coarsen_fraction = 0.1 # Fraction of low error that will coarsened
    max_h_level = 3 # Max number of refinements used, starting from initial mesh (before uniform refinement)
  [../]
[]


Thanks,
Bala

**Reply by Bala-1005** [reply]
Hello @jiangwen84. I did try to run the simulation with LU solver. The issue of non convergence when two solid parts of the domain meet still persists. There is no issue when the stiffness matrices are comparable.

Thanks,
Bala

**Reply by jiangwen84** [reply]
It is a bit difficult to tell what will be issue. Looks like the convergence issue might due to your physics, i.e. some quantities become unphysical, and the large material contrast trigger this issue? 

Anyway, could you attach your output? It might be useful to take a look at the linear and nonlinear iterations behavior.

**Reply by Bala-1005** [reply]
I was able to slightly solve the issue by increasing the number of nonlinear iterations to 100. But only for several timesteps. The issue returned with convergence being the issue. The output file is very large and therefore I am attaching the part where it starts to cut timesteps and stop converging.
[output.txt](https://github.com/idaholab/moose/files/6190636/output.txt)
[error.txt](https://github.com/idaholab/moose/files/6190639/error.txt)


Thanks,
Bala

**Reply by jiangwen84** [reply]
Try to set `solve_type = 'PJFNK'`

Maybe you want to loose your convergence tolerance a bit, i.e., nl_rel_tol = 1e-7 and nl_abs_tol=1e-7

**Reply by Bala-1005** [reply]
Thank you @jiangwen84 and everyone. I was able to solve the issue by increasing the nonlinear iterations number, loosening the tolerance a bit, and changing the boundary conditions from a displacement boundary condition to pressure boundary conditions. The simulation runs quite well now. 
One final problem, I hope :),  is that there are a few memory issues. Any suggestions on how to solve these issues? Here is the error file. I used --malloc-dump while running the file.

[el2b_KHS.txt](https://github.com/idaholab/moose/files/6213651/el2b_KHS.txt)

Thanks, 
Bala

**Reply by jiangwen84** [reply]
I am not sure what is happening. Is this two dimension problem? Seems the error comes from PETSc, but not sure about the reason.

**Reply by Bala-1005** [reply]
Yes. It is a two dimension problem. 
Thanks,
Bala
